File: telegramBot.py
import cloudscraper
import time
import bs4
import requests
import random
from datetime import datetime
import pytz
import json
import os
from telegram import Update
from telegram.ext import Application, CommandHandler, MessageHandler, filters, ContextTypes, JobQueue
import asyncio
import threading
import logging
from apscheduler.schedulers.asyncio import AsyncIOScheduler


# Set environment timezone to ensure apscheduler compatibility
os.environ["TZ"] = "Europe/Lisbon"
try:
    time.tzset()  # Update timezone (works on Unix/macOS)
except AttributeError:
    pass  # Windows doesn’t support tzset, rely on pytz

# Set up logging
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)
logger = logging.getLogger(__name__)

# Bot token (replace with your new secure token)
TOKEN = "7681224927:AAE96XaKmmRA8gJApv0kFxos6f36Js_4S7s"

# Directory to store JSON configs
CONFIG_DIR = "configs"
if not os.path.exists(CONFIG_DIR):
    os.makedirs(CONFIG_DIR)

# Global variable to control script running
running = False
script_thread = None

# Default timezone
TZ = pytz.timezone("Europe/Lisbon")

# Load existing configs
def load_configs():
    configs = []
    for filename in os.listdir(CONFIG_DIR):
        if filename.endswith(".json"):
            with open(os.path.join(CONFIG_DIR, filename), 'r') as f:
                config = json.load(f)
                configs.append(config)
    return configs

# Save config to JSON
def save_config(config):
    config_id = config.get("config_id", str(int(time.time() * 1000)))
    config["config_id"] = config_id
    with open(os.path.join(CONFIG_DIR, f"{config_id}.json"), 'w') as f:
        json.dump(config, f, indent=4)
    return config_id

# Delete config
def delete_config(config_id):
    config_path = os.path.join(CONFIG_DIR, f"{config_id}.json")
    if os.path.exists(config_path):
        os.remove(config_path)
        return True
    return False

# Main script function
def run_script(config):
    global running
    scraper = cloudscraper.create_scraper()
    base_url = config.get("base_url", "https://www.wethrift.com/api/submit-action")
    slug = config.get("slug", "dropship")
    min_sleep = config.get("min_sleep", 300)
    max_sleep = config.get("max_sleep", 1000)
    params_list = []

    for deal in config["deals"]:
        deal_id = deal["deal_id"]
        deal_code = deal["deal_code"]
        deal_position = deal.get("deal_position", "2")
        params_list.extend([
            {"slug": slug, "deal_id": deal_id, "type": "used", "value": "1", "deal_position": deal_position, "deal_code": deal_code},
            {"slug": slug, "deal_id": deal_id, "type": "code_working", "value": "yes", "deal_position": deal_position, "deal_code": deal_code}
        ])

    while running:
        logger.info("Sending API actions...")
        for params in params_list:
            if not running:
                break
            params["t"] = str(int(time.time() * 1000))
            timestamp = datetime.now(TZ).strftime("%Y-%m-%d %H:%M:%S")
            logger.info(f"Request sent at {timestamp}")
            try:
                response = scraper.post(base_url, params=params)
                logger.info(f"Status Code: {response.status_code}")
                logger.info(f"Response: {response.text}")
            except Exception as e:
                logger.error(f"Error: {e}")
            time.sleep(2)

        if not running:
            break
        sleep = random.uniform(min_sleep, max_sleep)
        logger.info(f"Waiting for {int(sleep)} seconds...")
        time.sleep(sleep)

# Middleware to ensure timezone-aware datetimes
async def timezone_middleware(update: Update, context: ContextTypes.DEFAULT_TYPE, next_handler):
    if update.message and update.message.date:
        update.message.date = update.message.date.replace(tzinfo=TZ)
    await next_handler(update, context)

# Start command
async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(
        "Welcome to the Script Management Bot!\n"
        "Available commands:\n"
        "/add - Add a new configuration\n"
        "/list - List all configurations\n"
        "/delete <config_id> - Delete a configuration\n"
        "/run <config_id> - Run a configuration\n"
        "/stop - Stop the running script\n"
        "/settime <config_id> <min_sleep> <max_sleep> - Set sleep time range"
    )

# Add new config
async def add(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(
        "Send the configuration in JSON format:\n"
        "{\n"
        '  "slug": "dropship",\n'
        '  "base_url": "https://www.wethrift.com/api/submit-action",\n'
        '  "min_sleep": 300,\n'
        '  "max_sleep": 1000,\n'
        '  "deals": [\n'
        '    {"deal_id": "P3P5D9X5JJ", "deal_code": "SAVE25", "deal_position": "2"},\n'
        '    {"deal_id": "ANOTHERID", "deal_code": "CODE2", "deal_position": "2"}\n'
        '  ]\n'
        "}"
    )
    context.user_data["awaiting_config"] = True

# Handle JSON config input
async def handle_message(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if context.user_data.get("awaiting_config"):
        try:
            config = json.loads(update.message.text)
            required_fields = ["deals"]
            for field in required_fields:
                if field not in config:
                    raise ValueError(f"Field '{field}' is required")
            for deal in config["deals"]:
                if "deal_id" not in deal or "deal_code" not in deal:
                    raise ValueError("Each deal must have 'deal_id' and 'deal_code'")
            config_id = save_config(config)
            context.user_data["awaiting_config"] = False
            await update.message.reply_text(f"Configuration saved with ID: {config_id}")
        except json.JSONDecodeError:
            await update.message.reply_text("Invalid JSON. Try again.")
        except ValueError as e:
            await update.message.reply_text(f"Error: {e}")
        except Exception as e:
            await update.message.reply_text(f"Error saving configuration: {e}")

# List configs
async def list_configs(update: Update, context: ContextTypes.DEFAULT_TYPE):
    configs = load_configs()
    if not configs:
        await update.message.reply_text("No configurations found.")
        return
    response = "Available configurations:\n"
    for config in configs:
        response += f"ID: {config['config_id']}\n"
        response += f"Deals: {len(config['deals'])}\n"
        response += f"Min Sleep: {config.get('min_sleep', 300)}s\n"
        response += f"Max Sleep: {config.get('max_sleep', 1000)}s\n"
        response += "---\n"
    await update.message.reply_text(response)

# Delete config
async def delete(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not context.args:
        await update.message.reply_text("Usage: /delete <config_id>")
        return
    config_id = context.args[0]
    if delete_config(config_id):
        await update.message.reply_text(f"Configuration {config_id} deleted.")
    else:
        await update.message.reply_text(f"Configuration {config_id} not found.")

# Run config
async def run(update: Update, context: ContextTypes.DEFAULT_TYPE):
    global running, script_thread
    if not context.args:
        await update.message.reply_text("Usage: /run <config_id>")
        return
    config_id = context.args[0]
    config_path = os.path.join(CONFIG_DIR, f"{config_id}.json")
    if not os.path.exists(config_path):
        await update.message.reply_text(f"Configuration {config_id} not found.")
        return
    if running:
        await update.message.reply_text("A script is already running. Use /stop to stop it.")
        return
    with open(config_path, 'r') as f:
        config = json.load(f)
    running = True
    script_thread = threading.Thread(target=run_script, args=(config,))
    script_thread.start()
    await update.message.reply_text(f"Running configuration {config_id}.")

# Stop script
async def stop(update: Update, context: ContextTypes.DEFAULT_TYPE):
    global running, script_thread
    if not running:
        await update.message.reply_text("No script is running.")
        return
    running = False
    if script_thread:
        script_thread.join()
        script_thread = None
    await update.message.reply_text("Script stopped.")

# Set sleep time
async def set_time(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if len(context.args) != 3:
        await update.message.reply_text("Usage: /settime <config_id> <min_sleep> <max_sleep>")
        return
    config_id, min_sleep, max_sleep = context.args
    try:
        min_sleep = float(min_sleep)
        max_sleep = float(max_sleep)
        if min_sleep < 0 or max_sleep < min_sleep:
            raise ValueError("Invalid values for min_sleep or max_sleep")
    except ValueError:
        await update.message.reply_text("min_sleep and max_sleep must be valid numbers, with min_sleep >= 0 and max_sleep >= min_sleep")
        return
    config_path = os.path.join(CONFIG_DIR, f"{config_id}.json")
    if not os.path.exists(config_path):
        await update.message.reply_text(f"Configuration {config_id} not found.")
        return
    with open(config_path, 'r') as f:
        config = json.load(f)
    config["min_sleep"] = min_sleep
    config["max_sleep"] = max_sleep
    save_config(config)
    await update.message.reply_text(f"Sleep time updated for configuration {config_id}: min_sleep={min_sleep}s, max_sleep={max_sleep}s")

def main():
    # Initialize JobQueue with pytz timezone
    job_queue = JobQueue()
    scheduler = AsyncIOScheduler(timezone=TZ)
    job_queue.set_scheduler(scheduler)

    # Initialize application with custom JobQueue
    app = Application.builder().token(TOKEN).job_queue(job_queue).build()
    
    # Add middleware to enforce timezone
    async def wrapped_handler(update: Update, context: ContextTypes.DEFAULT_TYPE):
        await timezone_middleware(update, context, lambda u, c: handler(u, c))
    
    # Register handlers
    app.add_handler(CommandHandler("start", lambda u, c: timezone_middleware(u, c, start)))
    app.add_handler(CommandHandler("add", lambda u, c: timezone_middleware(u, c, add)))
    app.add_handler(CommandHandler("list", lambda u, c: timezone_middleware(u, c, list_configs)))
    app.add_handler(CommandHandler("delete", lambda u, c: timezone_middleware(u, c, delete)))
    app.add_handler(CommandHandler("run", lambda u, c: timezone_middleware(u, c, run)))
    app.add_handler(CommandHandler("stop", lambda u, c: timezone_middleware(u, c, stop)))
    app.add_handler(CommandHandler("settime", lambda u, c: timezone_middleware(u, c, set_time)))
    app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, lambda u, c: timezone_middleware(u, c, handle_message)))

    app.run_polling()

if __name__ == "__main__":
    main()

File: telegramBot.js
const TelegramBot = require('node-telegram-bot-api');
const axios = require('axios');
const cloudscraper = require('cloudscraper');
const moment = require('moment-timezone');
const fs = require('fs-extra');
const path = require('path');

// Bot token
const TOKEN = '7681224927:AAE96XaKmmRA8gJApv0kFxos6f36Js_4S7s';
const bot = new TelegramBot(TOKEN, { polling: true });

// Configs directory
const CONFIG_DIR = path.join(__dirname, 'configs');
fs.ensureDirSync(CONFIG_DIR);

// Global state
let runningConfigs = new Set(); // Track running config IDs
let stopAllRequested = false; // Flag to stop all configs

// Timezone
const TZ = 'Europe/Lisbon';

// Load existing configs
function loadConfigs() {
  const configs = [];
  const files = fs.readdirSync(CONFIG_DIR).filter(file => file.endsWith('.json'));
  for (const file of files) {
    const config = fs.readJsonSync(path.join(CONFIG_DIR, file));
    configs.push(config);
  }
  return configs;
}

// Save config to JSON
function saveConfig(config) {
  const configId = config.config_id || Date.now().toString();
  config.config_id = configId;
  fs.writeJsonSync(path.join(CONFIG_DIR, `${configId}.json`), config, { spaces: 2 });
  return configId;
}

// Delete config
function deleteConfig(configId) {
  const configPath = path.join(CONFIG_DIR, `${configId}.json`);
  if (fs.existsSync(configPath)) {
    fs.removeSync(configPath);
    return true;
  }
  return false;
}

// Sleep function
const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

// Truncate message for Telegram
function truncateMessage(message, maxLength = 4000) {
  if (message.length <= maxLength) return message;
  return message.slice(0, maxLength - 3) + '...';
}

// Main script function
async function runScript(config, chatId) {
  const baseUrl = config.base_url || 'https://www.wethrift.com/api/submit-action';
  const slug = config.slug || 'dropship';
  const minSleep = config.min_sleep || 300;
  const maxSleep = config.max_sleep || 1000;
  const paramsList = [];
  const positionStatus = new Map(); // Track status by deal_position
  let consecutive403s = 0;
  const MAX_403S = 3; // Stop after 3 consecutive 403s

  // Initialize position status and build params for deal_positions 1-10
  for (let dealPosition = 1; dealPosition <= 10; dealPosition++) {
    positionStatus.set(dealPosition.toString(), { total: 0, successful: 0 });
    for (const deal of config.deals) {
      const dealId = deal.deal_id;
      const dealCode = deal.deal_code;
      positionStatus.get(dealPosition.toString()).total += 2; // Two requests per deal
      paramsList.push(
        { slug, deal_id: dealId, type: 'used', value: '1', deal_position: dealPosition.toString(), deal_code: dealCode, _deal_position: dealPosition.toString() },
        { slug, deal_id: dealId, type: 'code_working', value: 'yes', deal_position: dealPosition.toString(), deal_code: dealCode, _deal_position: dealPosition.toString() }
      );
    }
  }

  while (runningConfigs.has(config.config_id) && !stopAllRequested) {
    console.log(`[INFO] Sending API actions for config ${config.config_id}...`);
    let hasError = false;

    // Reset successful counts for this loop
    for (const [position] of positionStatus) {
      positionStatus.get(position).successful = 0;
    }

    for (const params of paramsList) {
      if (!runningConfigs.has(config.config_id) || stopAllRequested) break;
      params.t = Math.floor(Date.now()).toString();
      const timestamp = moment().tz(TZ).format('YYYY-MM-DD HH:mm:ss');
      console.log(`[INFO] Request sent at ${timestamp} for config ${config.config_id}, deal_position=${params._deal_position}, type=${params.type}`);
      try {
        const response = await cloudscraper({
          method: 'POST',
          url: baseUrl,
          qs: params,
          headers: {
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36',
            'Accept': 'application/json, text/plain, */*',
            'Accept-Language': 'en-US,en;q=0.9',
            'Referer': 'https://www.wethrift.com/',
            'Origin': 'https://www.wethrift.com',
            'X-Requested-With': 'XMLHttpRequest',
            'Connection': 'keep-alive',
          },
          jar: true, // Enable cookies
          challengesToSolve: 5, // Increase attempts to solve Cloudflare challenges
          resolveWithFullResponse: true, // Get full response object
        });
        console.log(`Status Code: ${response.statusCode}`);
        console.log(`Response: ${response.body}`);
        console.log('---');
        if (response.statusCode === 200 && JSON.parse(response.body).success) {
          positionStatus.get(params._deal_position).successful++;
        }
        consecutive403s = 0; // Reset counter on success
      } catch (error) {
        hasError = true;
        console.error(`[ERROR] Request failed: ${error.message}`);
        let errorMessage = `[ERROR] Request failed for config ${config.config_id}, deal_position=${params._deal_position}, type=${params.type}: ${error.message}\n`;
        if (error.response) {
          console.error(`Status Code: ${error.response.statusCode}`);
          console.error(`Response Body: ${error.response.body}`);
          console.error(`Headers: ${JSON.stringify(error.response.headers, null, 2)}`);
          errorMessage += `Status: ${error.response.statusCode}\nBody: ${truncateMessage(error.response.body, 1000)}\n`;
          if (error.response.statusCode === 403) {
            consecutive403s++;
            errorMessage += `Consecutive 403s: ${consecutive403s}/${MAX_403S}\n`;
            if (consecutive403s >= MAX_403S) {
              runningConfigs.delete(config.config_id);
              errorMessage += `Stopped config ${config.config_id} due to too many 403 errors. Check Cloudflare protection or deal parameters.`;
              await bot.sendMessage(config.chat_id || chatId, truncateMessage(errorMessage));
              return { success: false, error: config.config_id };
            }
          }
        } else if (error.request) {
          console.error(`Request Details: ${JSON.stringify(error.request, null, 2)}`);
          errorMessage += `Request failed without response.\n`;
        }
        console.error(`Stack: ${error.stack}`);
      }
      await sleep(2000); // 2 seconds between requests
    }

    // Send Telegram summary message
    if (runningConfigs.has(config.config_id) && !stopAllRequested) {
      let summaryMessage = `Request results for config ${config.config_id}:\n`;
      for (const [position, status] of positionStatus) {
        summaryMessage += `${status.successful}/${status.total} requests for position ${position} worked\n`;
      }
      try {
        await bot.sendMessage(config.chat_id || chatId, truncateMessage(summaryMessage));
      } catch (telegramError) {
        console.error(`[ERROR] Failed to send Telegram message: ${telegramError.message}`);
      }
    }

    if (!runningConfigs.has(config.config_id) || stopAllRequested) {
      return { success: !hasError, error: hasError ? config.config_id : null };
    }
    const sleepTime = Math.random() * (maxSleep - minSleep) + minSleep;
    console.log(`[INFO] Waiting for ${Math.round(sleepTime)} seconds for config ${config.config_id}...\n`);
    await sleep(sleepTime * 1000);
  }
  return { success: !hasError, error: hasError ? config.config_id : null };
}

// Start command
bot.onText(/\/start/, (msg) => {
  const chatId = msg.chat.id;
  bot.sendMessage(chatId, 
    'Welcome to the Script Management Bot!\n' +
    'Available commands:\n' +
    '/add - Add a new configuration\n' +
    '/list - List all configurations\n' +
    '/delete <config_id> - Delete a configuration\n' +
    '/run <config_id | all> - Run a configuration or all configurations\n' +
    '/stop <config_id | all> - Stop a configuration or all configurations\n' +
    '/settime <config_id> <min_sleep> <max_sleep> - Set sleep time range'
  );
});

// Add new config
bot.onText(/\/add/, (msg) => {
  const chatId = msg.chat.id;
  bot.sendMessage(chatId, 
    'Send the configuration in JSON format:\n' +
    '{\n' +
    '  "slug": "dropship",\n' +
    '  "base_url": "https://www.wethrift.com/api/submit-action",\n' +
    '  "min_sleep": 300,\n' +
    '  "max_sleep": 1000,\n' +
    '  "deals": [\n' +
    '    {"deal_id": "P3P5D9X5JJ", "deal_code": "SAVE25", "deal_position": "1"},\n' +
    '    {"deal_id": "P3P5D9X5JJ", "deal_code": "SAVE25", "deal_position": "2"}\n' +
    '  ]\n' +
    '}'
  );
  bot.once('message', (nextMsg) => {
    if (nextMsg.chat.id !== chatId) return;
    try {
      const config = JSON.parse(nextMsg.text);
      if (!config.deals) throw new Error("Field 'deals' is required");
      for (const deal of config.deals) {
        if (!deal.deal_id || !deal.deal_code || !deal.deal_position) {
          throw new Error("Each deal must have 'deal_id', 'deal_code', and 'deal_position'");
        }
      }
      config.chat_id = chatId; // Store chat_id for error notifications
      const configId = saveConfig(config);
      bot.sendMessage(chatId, `Configuration saved with ID: ${configId}`);
    } catch (error) {
      bot.sendMessage(chatId, `Error: ${error.message}`);
    }
  });
});

// List configs
bot.onText(/\/list/, (msg) => {
  const chatId = msg.chat.id;
  const configs = loadConfigs();
  if (!configs.length) {
    bot.sendMessage(chatId, 'No configurations found.');
    return;
  }
  let response = 'Available configurations:\n';
  for (const config of configs) {
    response += `ID: ${config.config_id}\n`;
    response += `Deals: ${config.deals.length}\n`;
    response += `Min Sleep: ${config.min_sleep || 300}s\n`;
    response += `Max Sleep: ${config.max_sleep || 1000}s\n`;
    response += '---\n';
  }
  bot.sendMessage(chatId, response);
});

// Delete config
bot.onText(/\/delete (.+)/, (msg, match) => {
  const chatId = msg.chat.id;
  const configId = match[1];
  if (deleteConfig(configId)) {
    bot.sendMessage(chatId, `Configuration ${configId} deleted.`);
  } else {
    bot.sendMessage(chatId, `Configuration ${configId} not found.`);
  }
});

// Run config
bot.onText(/\/run (.+)/, async (msg, match) => {
  const chatId = msg.chat.id;
  const configId = match[1];

  if (configId === 'all') {
    const configs = loadConfigs();
    if (!configs.length) {
      bot.sendMessage(chatId, 'No configurations found.');
      return;
    }
    if (runningConfigs.size > 0) {
      bot.sendMessage(chatId, 'One or more configs are already running. Use /stop to stop them.');
      return;
    }
    bot.sendMessage(chatId, `Starting configs for all (${configs.length} configs).`);
    while (!stopAllRequested) {
      let successful = 0;
      let errors = 0;
      const errorConfigIds = [];
      const promises = configs.map(config => {
        if (runningConfigs.has(config.config_id)) return Promise.resolve({ success: false, error: config.config_id });
        config.chat_id = chatId; // Store chat_id for notifications
        runningConfigs.add(config.config_id);
        return runScript(config, chatId).catch(error => {
          console.error(`[ERROR] Script error for config ${config.config_id}: ${error.message}`);
          return { success: false, error: config.config_id };
        }).finally(() => {
          runningConfigs.delete(config.config_id);
        });
      });
      const results = await Promise.all(promises);
      results.forEach(result => {
        if (result.success) successful++;
        else {
          errors++;
          if (result.error) errorConfigIds.push(result.error);
        }
      });
      if (stopAllRequested) break;
      const summaryMessage = `Summary: Success for ${successful}/${configs.length} configs, errors = ${errors};${errorConfigIds.join(',') || 'none'}`;
      try {
        await bot.sendMessage(chatId, truncateMessage(summaryMessage));
      } catch (telegramError) {
        console.error(`[ERROR] Failed to send Telegram summary: ${telegramError.message}`);
      }
      const globalSleep = Math.random() * (1000 - 300) + 300; // Global sleep between 300-1000s
      console.log(`[INFO] Waiting for ${Math.round(globalSleep)} seconds before next iteration...`);
      await sleep(globalSleep * 1000);
    }
    return;
  }

  const configPath = path.join(CONFIG_DIR, `${configId}.json`);
  if (!fs.existsSync(configPath)) {
    bot.sendMessage(chatId, `Configuration ${configId} not found.`);
    return;
  }
  if (runningConfigs.has(configId)) {
    bot.sendMessage(chatId, `Configuration ${configId} is already running.`);
    return;
  }
  const config = fs.readJsonSync(configPath);
  config.chat_id = chatId; // Store chat_id for error notifications
  runningConfigs.add(config.config_id);
  bot.sendMessage(chatId, `Running configuration ${configId}.`);
  runScript(config, chatId).catch(error => {
    console.error(`[ERROR] Script error: ${error.message}`);
    runningConfigs.delete(config.config_id);
    bot.sendMessage(chatId, `Script stopped due to error: ${error.message}`);
  }).finally(() => {
    runningConfigs.delete(config.config_id);
  });
});

// Stop script
bot.onText(/\/stop (.+)/, (msg, match) => {
  const chatId = msg.chat.id;
  const configId = match[1];

  if (configId === 'all') {
    if (runningConfigs.size === 0) {
      bot.sendMessage(chatId, 'No configs are running.');
      return;
    }
    stopAllRequested = true;
    runningConfigs.clear();
    bot.sendMessage(chatId, 'Stopped all configs.');
    stopAllRequested = false; // Reset flag
    return;
  }

  const configPath = path.join(CONFIG_DIR, `${configId}.json`);
  if (!fs.existsSync(configPath)) {
    bot.sendMessage(chatId, `Configuration ${configId} not found.`);
    return;
  }
  if (!runningConfigs.has(configId)) {
    bot.sendMessage(chatId, `Configuration ${configId} is not running.`);
    return;
  }
  runningConfigs.delete(configId);
  bot.sendMessage(chatId, `Script for configuration ${configId} stopped.`);
});

// Set sleep time
bot.onText(/\/settime (.+) (.+) (.+)/, (msg, match) => {
  const chatId = msg.chat.id;
  const [configId, minSleep, maxSleep] = match.slice(1);
  const configPath = path.join(CONFIG_DIR, `${configId}.json`);
  if (!fs.existsSync(configPath)) {
    bot.sendMessage(chatId, `Configuration ${configId} not found.`);
    return;
  }
  try {
    const min = parseFloat(minSleep);
    const max = parseFloat(maxSleep);
    if (min < 0 || max < min) {
      throw new Error('Invalid values for min_sleep or max_sleep');
    }
    const config = fs.readJsonSync(configPath);
    config.min_sleep = min;
    config.max_sleep = max;
    saveConfig(config);
    bot.sendMessage(chatId, `Sleep time updated for configuration ${configId}: min_sleep=${min}s, max_sleep=${max}s`);
  } catch (error) {
    bot.sendMessage(chatId, `Error: ${error.message}`);
  }
});

console.log('Bot is running...');